openmmtools/cache.py

```diff
@@ -0,0 +1,352 @@
+    LRUCache.
+
+    >>> context_cache = ContextCache(capacity=1, time_to_live=5)
+    >>> context2, integrator2 = context_cache.get_context(thermodynamic_state,
+    ...                                                   openmm.VerletIntegrator(time_step))
+    >>> context3, integrator3 = context_cache.get_context(thermodynamic_state_npt,
+    ...                                                   openmm.VerletIntegrator(time_step))
+    >>> len(context_cache)
+    1
+
+    See Also
+    --------
+    LRUCache
+    states.ThermodynamicState.is_state_compatible
+
+    """
+
+    def __init__(self, **kwargs):
+        self._lru = LRUCache(**kwargs)
+
+    def __len__(self):
+        return len(self._lru)
+
+    def get_context(self, thermodynamic_state, integrator, platform=None):
+        """Return a context in the given thermodynamic state.
+
+        In general, the Context must be considered newly initialized. This
+        means that positions and velocities must be set afterwards.
+
+        This creates a new Context if no compatible one has been cached.
+        If a compatible Context exists, the ThermodynamicState is applied
+        to it, and the Context integrator state is changed to match the
+        one passed as an argument. As a consequence, the returned integrator
+        is guaranteed to be in the same state as the one provided, but it
+        can be a different instance. This is to minimize the number of
+        Contexts objects cached that use the same or very similar integrator.
+
+        Parameters
+        ----------
+        thermodynamic_state : states.ThermodynamicState
+            The thermodynamic state of the system.
+        integrator : simtk.openmm.Integrator
+            The integrator for the context.
+        platform : simtk.openmm.Platform, optional
+            The OpenMM platform to use. If None, OpenMM tries to select
+            the fastest one available (default is None).
+
+        Returns
+        -------
+        context : simtk.openmm.Context
+            The context in the given thermodynamic system.
+        context_integrator : simtk.openmm.Integrator
+            The integrator to be used to propagate the Context. Can be
+            a difference instance from the one passed as an argument.
+
+        """
+        context_id = self._generate_context_id(thermodynamic_state, integrator,
+                                               platform)
+        try:
+            context = self._lru[context_id]
+        except KeyError:
+            context = thermodynamic_state.create_context(integrator, platform)
+            self._lru[context_id] = context
+        context_integrator = context.getIntegrator()
+
+        # Update state of system and integrator of the cached context.
+        self._copy_integrator_state(integrator, context_integrator)
+        thermodynamic_state.apply_to_context(context)
+        return context, context_integrator
```

Comment:
Two integrators are compatible if they differ by only few parameters (e.g. step size, temperature). In this case `ContextCache` recycle the same `Context` and simply set the parameters of its `Integrator` to the ones requested.

This is why I made `get_context()` return a tuple `(Context, Integrator)` instead of just the `Context`. I thought this would make it clear (on top of the documentation) that the returned integrator must be used.